source/common/router/req_header_formatter.h

```diff
@@ -0,0 +1,100 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include "common/common/logger.h"
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Interface for all types of header formatters used for custom request headers.
+ */
+class HeaderFormatter {
+
+public:
+  virtual ~HeaderFormatter(){};
+
+  virtual std::string format(const Envoy::Http::AccessLog::RequestInfo& request_info) const PURE;
+};
+
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterSharedPtr;
```

Comment:
@RomanDzhabarov I tried moving this to a unique_ptr and also tried using std::move inside the parse method in RequestHeaderPrarser but it still looks like the compiler is not happy with it. Once i change it to a shared_ptr it starts working.